libkbfs/md_ops.go

```diff
@@ -53,14 +53,270 @@ func (md *MDOpsStandard) convertVerifyingKeyError(ctx context.Context,
+	default:
+		return false, err
+	}
+
+	md.log.CDebugf(ctx, "Revision %d for %s was signed by a device that was "+
+		"revoked at time=%d,root=%d; checking via Merkle",
+		irmd.Revision(), irmd.TlfID(), info.Time, info.MerkleRoot.Seqno)
+	ctx = context.WithValue(ctx, ctxMDOpsSkipKeyVerification, struct{}{})
+
+	kbfsRoot, merkleNodes, rootSeqno, _, err :=
+		md.config.MDServer().FindNextMD(ctx, rmds.MD.TlfID(),
+			info.MerkleRoot.Seqno)
+	if err != nil {
+		return false, err
+	}
+	if len(merkleNodes) == 0 {
+		// This can happen legitimately if we are still inside the
+		// error window and no new merkle trees have been made yet, or
+		// the server could be lying to us.
+		md.log.CDebugf(ctx, "The server claims there haven't been any "+
+			"KBFS merkle trees published since the revocation")
+		// Verify the chain up to the current head.  By using `ctx`,
+		// we'll avoid infinite loops in the writer-key-checking code
+		// by skipping revoked key verification.  This is ok, because
+		// we only care about the hash chain for the purposes of
+		// verifying `irmd`.
+		chain, err := getMergedMDUpdates(
+			ctx, md.config, irmd.TlfID(), irmd.Revision()+1, nil)
+		if err != nil {
+			return false, err
+		}
+		if len(chain) > 0 {
+			err = irmd.CheckValidSuccessor(
+				irmd.mdID, chain[0].ReadOnlyRootMetadata)
+			if err != nil {
+				return false, err
+			}
+		}
+
+		// TODO(KBFS-2956): check the most recent global merkle root
+		// and KBFS merkle root ctimes and make sure they fall within
+		// the expected error window with respect to the revocation.
+		// Also eventually check the blockchain-published merkles to
+		// make sure the server isn't lying (though that will have a
+		// much larger error window).
+		return true, nil
+	}
+
+	md.log.CDebugf(ctx,
+		"Next KBFS merkle root is %d, included in global merkle root seqno=%d",
+		kbfsRoot.SeqNo, rootSeqno)
+
+	// Decode (and possibly decrypt) the leaf node, so we can see what
+	// the given MD revision number was for the MD that followed the
+	// revoke.
+	leaf, err := md.makeMerkleLeaf(
+		ctx, irmd.ReadOnlyRootMetadata, kbfsRoot,
+		merkleNodes[len(merkleNodes)-1])
+	if err != nil {
+		return false, err
+	}
+
+	// If the given revision comes after the merkle leaf revision,
+	// then don't verify it.
+	if irmd.Revision() > leaf.Revision {
+		return false, MDWrittenAfterRevokeError{
+			irmd.TlfID(), irmd.Revision(), leaf.Revision, verifyingKey}
+	} else if irmd.Revision() == leaf.Revision {
+		return true, nil
+	}
+
+	// Otherwise it's valid, as long as there's a valid chain of MD
+	// revisions between the two, and the global root info checks out
+	// as well.  By using `ctx`, we'll avoid infinite loops in the
+	// writer-key-checking code by skipping revoked key verification.
+	// This is ok, because we only care about the hash chain for the
+	// purposes of verifying `irmd`.
+	chain, err := getMergedMDUpdatesWithEnd(
+		ctx, md.config, irmd.TlfID(), irmd.Revision()+1, leaf.Revision, nil)
```

Comment:
In the `decryptMerkleLeaf` case, we would have already retrieved this chain (via repeated iterations of `fetchLoop`). I wonder if we should return that to minimize repeated work.

Reply:
Maybe we would have already fetched it, but maybe not.  They could be fetching different ranges.  It's possible `decryptMerkleLeaf` could have already verified part of the chain, but it could be only a small part of it, depending on what `leaf.Revision` is.  In any case, I'm specifically using `md_util` functions here, rather than calling `mdOps.getRange()` directly, because I want it to get MDs from the cache whenever possible.  So it should already avoid re-fetching and re-verifying any MDs more than once, as long as they were readable and cacheable.

Reply:
Cool, thanks for the explanation.